etcdlock/rwmutex2.go

```diff
@@ -31,16 +29,12 @@ func (rwm *RWMutex) RLock(ctx context.Context) error {
 		return err
 	}
 	rwm.myKey = rk
-	// wait until nodes with "write-" and a lower revision number than myKey are gone
-	for {
-		if done, werr := rwm.waitOnLastRev(ctx, rwm.pfx+"write"); done || werr != nil {
-			err = rwm.myKey.Delete()
-			if err != nil {
-				return err
-			}
-			return werr
-		}
+	// wait until nodes with "/write" and a lower revision number than myKey are gone
+	if err = rwm.waitOnLastRev(ctx, rwm.pfx+"write"); err != nil {
+		_ = rwm.myKey.Delete()
```

Comment:
Why did you ignore the possible Delete error here?

Reply:
if RLock or RWLock returns the error from Delete function the real reason of "why was not possible to get a lock" will be masked.

A type of error the function waitOnLastRev can return is "context deadline exceeded" as this is an important error message it should not be overwritten.

Maybe concatenating the errors' messages can be a solution. What do you think?